airbytehq/airbyte
# Introduce config injection

## TODO

* Tests
* Check with @jdpgrailsdev whether this is the right approach in the first place

## What

Fixes #22439
Fixes #22440

Adds actor definition config injection table that injects a json blob under a specified injection path into the connection configration of source or destination before running a sync/check/discover job.

I slightly deviated from the description in by avoiding a UUID `id` and instead going with a composite PK over injection_path and actor_definition_id. Semantically it doesn't make sense to have more than one config injection that targets the same injection path and actor definition id and when writing them as part of #22457 injection_path and actor_definition_id will be the identifying properties as well - introducing a synthetic UUID will only complicate things in this specific case, especially since these objects are not planned to be exposed via API at any point.

## How

* Add a new table and models as usual.
* Introduce `ConfigInjector` helper class that fetches and injects blobs into the config and make it known to DI for worker and server.
* Call the helper in all the places a job config is put together:
  * `airbyte-commons-server/src/main/java/io/airbyte/commons/server/scheduler/DefaultSynchronousSchedulerClient.java`
  * `airbyte-persistence/job-persistence/src/main/java/io/airbyte/persistence/job/factory/DefaultSyncJobFactory.java`
  * `airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/GenerateInputActivityImpl.java`
* The write method is not used yet but will be used in a follow-up PR
* Adds tests for new methods in config repository as well as config injector
  * I only introduced a single test suite for both parts together because they are fairly slim already and I don't see how we gain much by having separate tests
* Extends existing tests for the places where job configs are put together

## How to test

To test this end-to-end, set up a connection with the pokeapi source, then execute the following query:
```
insert into actor_definition_config_injection values ('"pikachu"', 'pokemon_name', '6371b14b-bc68-4236-bfbd-468e8df8e968');
```

This will cause all sync/discover/check jobs to run for pikachu, no matter what's actually configured. Setting to an invalid value should fail the calls with the expected error message.
